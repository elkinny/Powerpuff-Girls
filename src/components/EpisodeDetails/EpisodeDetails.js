import React from 'react';
import { useSelector } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faFilm } from '@fortawesome/free-solid-svg-icons';

import { viewedEpisodeSelector } from '../../store/selectors';
import InfoCard from '../InfoCard';

const EpisodeDetails = () => {
  const currentEpisode = useSelector(viewedEpisodeSelector);

  return (
    <InfoCard
      image={currentEpisode.image ? currentEpisode.image.original : '/placeholder.png'}
      name={currentEpisode.name}
      summary={currentEpisode.summary}
      subTitle={(
        <>
          <FontAwesomeIcon icon={faFilm} />
          {`${currentEpisode.season}.${currentEpisode.number}`}
        </>
      )}
      link={{ url: '/', text: 'Back' }}
      isColumn
    />
  );
};

EpisodeDetails.getProps = async (context) => {
  const res = await fetch(`${process.env.API_URL}/episodes/${context.params.episodeId}`);

  const data = await res.json();

  if (!data) {
    return {
      props: { initialReduxState: { viewedEpisode: {} } },
    };
  }

  return {
    props: {
      initialReduxState: {
        viewedEpisode: data,
      },
    },
  };
};

export default EpisodeDetails;
