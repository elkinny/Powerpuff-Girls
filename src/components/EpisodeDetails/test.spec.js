import React from 'react';
import { render } from '@testing-library/react';

import EpisodeDetails from './EpisodeDetails';

jest.mock('../InfoCard', () => () => 'InfoCard');
jest.mock('react-redux', () => ({
  useSelector: () => ({}),
}));
global.fetch = jest.fn(() => Promise.resolve({
  json: () => Promise.resolve({ }),
}));

describe('EpisodeDetails', () => {
  test('renders EpisodeDetails component', () => {
    const { container } = render(<EpisodeDetails />);

    expect(container).toMatchSnapshot();
  });
});

describe('getProps', () => {
  beforeEach(() => {
    fetch.mockClear();
  });

  test('return formed data', async () => {
    fetch.mockImplementationOnce(() => Promise.resolve({
      json: () => Promise.resolve({ viewedEpisode: 1 }),
    }));
    const context = { params: { episodeId: 1 } };

    const result = await EpisodeDetails.getProps(context);

    expect(result).toMatchSnapshot();
  });

  test('return formed data when data is empty', async () => {
    fetch.mockImplementationOnce(() => Promise.resolve({
      json: () => Promise.resolve(null),
    }));
    const context = { params: { episodeId: 1 } };

    const result = await EpisodeDetails.getProps(context);

    expect(result).toMatchSnapshot();
  });
});
