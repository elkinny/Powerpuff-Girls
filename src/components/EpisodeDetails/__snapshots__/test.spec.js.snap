// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodeDetails renders EpisodeDetails component 1`] = `
<div>
  InfoCard
</div>
`;

exports[`getProps return formed data 1`] = `
Object {
  "props": Object {
    "initialReduxState": Object {
      "episodes": Object {
        "viewedEpisode": 1,
      },
    },
  },
}
`;

exports[`getProps return formed data when data is empty 1`] = `
Object {
  "props": Object {
    "initialReduxState": Object {
      "episodes": Object {},
    },
  },
}
`;
